lib/datalog/access.go

```
@@ -0,0 +1,366 @@
+	reverseMappings map[uint32]string
+	errorStrings    map[int]string
+}
+
+const (
+	loginTraitHash = 0
+	keyJoin        = "_"
+	userIndex      = 0
+	loginIndex     = 1
+	nodeIndex      = 2
+	roleIndex      = 3
+)
+
+// QueryNodeAccess returns a list of accesses to Teleport.
+func QueryNodeAccess(ctx context.Context, client auth.ClientI, req NodeAccessRequest) (*NodeAccessResponse, error) {
+	resp := NewNodeAccessResponse()
+	resp.addToMap(types.Wildcard)
+
+	// Since we can filter on specific usernames, this conditional will only retrieve and build the facts for the required user(s).
+	if req.Username != "" {
+		u, err := client.GetUser(req.Username, false)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		resp.createUserPredicates(u, req.Login)
+	} else {
+		us, err := client.GetUsers(false)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		for _, u := range us {
+			resp.createUserPredicates(u, req.Login)
+		}
+	}
+
+	rs, err := client.GetRoles(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	for _, r := range rs {
+		resp.createRolePredicates(r)
+	}
+
+	ns, err := client.GetNodes(ctx, req.Namespace)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	for _, n := range ns {
+		if len(req.Node) == 0 || n.GetHostname() == req.Node {
+			resp.createNodePredicates(n)
+		}
+	}
+
+	b, err := resp.facts.Marshal()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	ptr := (*C.uchar)(C.CBytes(b))
+	defer C.free(unsafe.Pointer(ptr))
+	res := (*C.uchar)(C.CBytes(make([]byte, len(b))))
+	l := C.process_access(ptr, res, C.size_t(len(b)), C.size_t(len(b)))
+	if int(l) < 0 {
+		return nil, trace.Wrap(fmt.Errorf(resp.errorStrings[int(l)]))
+	}
+
+	// Couldn't write to result since there is not enough memory allocated
+	if int(l) > len(b) {
+		res = (*C.uchar)(C.CBytes(make([]byte, l)))
+		l = C.process_access(ptr, res, C.size_t(len(b)), C.size_t(l))
+		if int(l) < 0 {
+			return nil, trace.Wrap(fmt.Errorf(resp.errorStrings[int(l)]))
```

Usually when you want to return an error string, we do `trace.BadParameter("error message")` so you could do that instead of wrapping fmt.Errorf.